packages/@aws-cdk/aws-amplify/lib/app.ts

```diff
@@ -0,0 +1,287 @@
+  /**
+   * App Default Domain
+   *
+   * @attribute
+   */
+  public readonly appDefaultDomain: string;
+
+  /**
+   * Resource
+   */
+  private readonly resource: CfnApp;
+
+  private readonly basicAuthResolver: BasicAuthResolver = new BasicAuthResolver();
+
+  private readonly environmentVariablesResolver: EnvironmentVariablesResolver = new EnvironmentVariablesResolver();
+
+  constructor(scope: Construct, id: string, props: AppProps) {
+    super(scope, id, props);
+
+    if (props.basicAuthConfig) {
+      this.setBasicAuth(props.basicAuthConfig.username, props.basicAuthConfig.password);
+    }
+
+    if (props.environmentVariables && props.environmentVariables.length > 0) {
+      this.environmentVariablesResolver.addEnvironmentVariables(...props.environmentVariables);
+    }
+
+    const resource = new CfnApp(scope, 'App', {
+      accessToken: props.accessToken,
+      basicAuthConfig: this.basicAuthResolver,
+      buildSpec: props.buildSpec,
+      customRules: props.customRules,
+      description: props.description,
+      environmentVariables: this.environmentVariablesResolver,
+      name: props.name,
+      oauthToken: props.oauthToken,
+      repository: props.repository,
+      tags: props.tags
+    });
+
+    if (props.iamServiceRole) {
+      resource.iamServiceRole = props.iamServiceRole.roleArn;
+    }
+
+    if (props.buildSpec && !props.iamServiceRole) {
+      // create lazy role
+      const role = new LazyRole(scope, 'lazy-role', {
+        assumedBy: new ServicePrincipal('amplify.amazonaws.com')
+      });
+
+      resource.iamServiceRole = role.roleArn;
+    }
+
+    this.appArn = resource.attrArn,
+    this.appDefaultDomain = resource.attrDefaultDomain;
+    this.appId = resource.attrAppId;
+    this.appName = resource.attrAppName;
+
+    this.resource = resource;
+  }
+
+  public validate(): string[] {
+    const errors: string[] = [];
+
+    if (this.resource.buildSpec && !this.resource.iamServiceRole) {
+      errors.push(`You need to specify a service role when using buildspecs`);
+    }
+
+    return errors;
+  }
+
+  /**
+   * Add Service Role
+   *
+   * @param role
+   */
+  public addServiceRole(role: IRole) {
+    this.resource.iamServiceRole = role.roleArn;
+  }
+
+  /**
+   * Set Basic Auth on App
+   *
+   * @param username
+   * @param password
+   */
+  public setBasicAuth(username: string, password: string) {
+    this.basicAuthResolver.basicAuthConfig({
+      enableBasicAuth: true,
+      password,
+      username
+    });
+  }
+
+  /**
+   * Add Environment Variable to App
+   *
+   * @param name
+   * @param value
+   */
+  public addEnvironmentVariable(name: string, value: string) {
+    this.environmentVariablesResolver.addEnvironmentVariable(name, value);
+  }
+}
+
+/**
+ * App Interface
+ */
+export interface IApp extends IResource {
+  /**
+   * App ID
+   *
+   * @attribute
+   */
+  readonly appId: string;
+}
+
+/**
+ * App Options
+ */
+export interface AppOptions extends ResourceProps {
```

Comment:
We usually don’t extend ResourceProps. Physical name
Should be a resource-specific prop (eg appName).

Reply:
Same for Branch